### PR TITLE
Cron Support for CDC

### PR DESCRIPTION
**Pull request by Ajanthy**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Related Issues
https://github.com/wso2/streaming-integrator/issues/115
https://github.com/siddhi-io/siddhi-io-cdc/issues/58

### CODE REVIEW

File: component/pom.xml [hunk 1]
```diff
@@ -585,6 +585,7 @@
                             io.siddhi.query.*;version="${siddhi.import.version.range}",
                             *;resolution:=optional
                         </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (ramindu90) [inline component/pom.xml:588]: Avoid using dynamic imports. Add the necessary packages in the import-packages section

File: component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java [hunk 2]
```diff
@@ -252,6 +258,15 @@
                         type = DataType.INT,
                         optional = true,
                         defaultValue = "-1"
+                ),
+                @Parameter(
+                        name = "cron.expression",
+                        description = "This is used to specify a timestamp in cron expression. " +
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java:264]: I think the description is wrong. Please check

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCCronExecutor.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            //Trigger the job to at given cron expression
+            Trigger trigger = TriggerBuilder
+                    .newTrigger()
+                    .withIdentity(CDCSourceConstants.TRIGGER_NAME, CDCSourceConstants.TRIGGER_GROUP)
+                    .withSchedule(
+                            CronScheduleBuilder.cronSchedule(cronExpression)).build();
+            // Tell quartz to schedule the job using our trigger
+            scheduler.scheduleJob(cron, trigger);
+        } catch (SchedulerException e) {
+            log.error("The error occurs at scheduler start in SiddhiApp : " + e);
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCCronExecutor.java:80]: Shall we change this as "An error occurred when scheduling job in siddhi app for cdc polling"

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java [hunk 4]
```diff
@@ -185,6 +190,9 @@ public void resume() {
     public void run() {
         try {
             pollingStrategy.poll();
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java:192]: Shouldn't this come in the else part? In this scenario, whenever the cdc source starts running, a poll() happens right. Please confirm

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java [hunk 5]
```diff
@@ -76,52 +107,41 @@ public void poll() {
...
+                for (int i = 1; i <= metadata.getColumnCount(); i++) {
+                    String key = metadata.getColumnName(i);
+                    Object value = resultSet.getObject(key);
+                    detailsMap.put(key.toLowerCase(Locale.ENGLISH), value);
                 }
+                lastReadPollingColumnValue = resultSet.getString(pollingColumn);
+                handleEvent(detailsMap);
             }
         } catch (SQLException ex) {
-            throw new CDCPollingModeException(buildError("Error in polling for changes on %s.", tableName), ex);
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:122]: Any reason to log the error without throwing?

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java [hunk 6]
```diff
@@ -76,52 +107,41 @@ public void poll() {
...
+            log.error(buildError("Error occurred while processing records in table %s.", tableName), ex);
         } finally {
-            CDCPollingUtil.cleanupConnection(resultSet, statement, connection);
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:124]: We should close connection and statement after using. And re-initialize when using again.

File: component/src/main/java/io/siddhi/extension/io/cdc/util/CDCSourceConstants.java [hunk 7]
```diff
@@ -91,4 +91,12 @@
     public static final String MONGO_SET = "$set";
     public static final String MONGO_OBJECT_NUMBER_LONG = "$numberLong";
     public static final String MONGO_OBJECT_NUMBER_DECIMAL = "$numberDecimal";
+
+    // Constants for Cron Support
+    public static final String CRON_EXPRESSION = "cron.expression";
+    public static final String JOB_GROUP = "JobGroup";
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/util/CDCSourceConstants.java:97]: Shall we add some meaningful names if those are getting printed when debugging or on errors?? Eg: CDCPollingGroup

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java [hunk 8]
```diff
@@ -92,7 +97,7 @@ public CDCPoller(String url, String username, String password, String tableName,
         } else {
             log.debug(DefaultPollingStrategy.class + " is selected as the polling strategy.");
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java:98]: Shall we validate and throw app creation error if cron and some other strategy is selected?

File: component/pom.xml (hunk 1 above)
Comment (Ajanthy) [reply component/pom.xml]: fixed in ee838f87eb853dd4cccb33250c2fa8f7d620a3cc

File: component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java (hunk 2 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java]: fixed in ee838f87eb853dd4cccb33250c2fa8f7d620a3cc

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCCronExecutor.java (hunk 3 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCCronExecutor.java]: fixed in ee838f87eb853dd4cccb33250c2fa8f7d620a3cc

File: component/src/main/java/io/siddhi/extension/io/cdc/util/CDCSourceConstants.java (hunk 7 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/util/CDCSourceConstants.java]: fixed in ee838f87eb853dd4cccb33250c2fa8f7d620a3cc

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java (hunk 8 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java:98]: fixed in ee838f87eb853dd4cccb33250c2fa8f7d620a3cc

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java (hunk 5 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:122]: Followed the same approach.

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java (hunk 6 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:124]: Connection and statement are closed at the end of the method.

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java (hunk 4 above)
Comment (Ajanthy) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/CDCPoller.java:192]: `poll()` is called to set the lastReadPollingColumnValue. If the `poll()` is called in the `scheduleJob()` then records which has been inserted between the time app starts running and the first cronschedule will be missed. To ignore this issue the lastReadPollingColumnValue has to be stored. So it shouldn't come in the else part.

**Pull request merged**